lib/shoryuken/processor.rb

```diff
@@ -2,8 +2,77 @@ module Shoryuken
   class Processor
     include Util
 
-    def initialize(manager)
-      @manager = manager
+    BATCH_LIMIT = 10
+
+    def initialize(fetcher, polling_strategy)
+      @fetcher = fetcher
+      @polling_strategy = polling_strategy
+      # what is the worst thing that could happen because it's not atomic?
+      # is it possible that it evaluates to true even if we never set it to true?
+      @done = false
+    end
+
+    def start
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      # can it throw exception?
+      # what should we do if it dies?
+      @thread ||= spawn_thread(&method(:loop))
+    end
+
+    def terminate(wait=false)
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.value if wait
+    end
+
+    def kill
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.raise ::Shoryuken::Shutdown
+    end
+
+    def alive?
+      # is this ok?
+      !!@thread.status
+    end
+
+    private
+
+    def loop
+      while !@done
+        do_work
+      end
+    rescue ::Shoryuken::Shutdown
+      # killed, bail!
+    rescue => e
+      # are we safe to retry here? What are kind of errors we do not want to recover from?
+      # OOM?
+      logger.debug { "Processor caught exception: #{e}" }
+      retry
+    end
+
+    def do_work
+      queue = @polling_strategy.next_queue
+      return logger.info { 'Doing nothing, all queues are paused' } if queue.nil?
+      batched_queue?(queue) ? dispatch_batch(queue) : dispatch_single_message(queue)
+    end
+
+    def dispatch_batch(queue)
+      batch = @fetcher.fetch(queue, BATCH_LIMIT)
+      @polling_strategy.messages_found(queue.name, batch.size)
+      process(queue.name, patch_batch!(batch))
+    end
+
+    def dispatch_single_message(queue)
+      messages = @fetcher.fetch(queue, 1)
+      @polling_strategy.messages_found(queue.name, messages.size)
+      process(queue.name, messages.first) if messages.length > 0
+    end
```

Comment:
It is not possible with the current set of interfaces (all queues are paused so we get nil), but our interfaces are not written in stone.

It's good that you mentioned long polling - since processors are independent, it should be totally ok to use long-polling all the time. Ish. If we have just one queue, we should use long polling for sure.
If we have multiple queues, we could not want to block for too long on q1 if there is q2 being full. I am wondering if long polling wait time is the actual parameter we want to tweak, instead of the heartbeat interval we've discussed in https://github.com/phstc/shoryuken/pull/291. I guess being able to configure `queue_empty_pause` for an empty queue and `long_polling_wait_time` will be sufficient for all kinds of workloads.

Just to recap what would that approach bring us:
- optimal performance for "one queue" configurations, regardless concurrency level
- configurable tradeoff for cost vs throughput via `long_polling_wait_time` and `queue_empty_pause` parameters. People could optimise the number of calls to SQS or throughput of the system, depending on their needs.